### ISSUE TITLE
[question] How to use a lockfile to override requirements versions?

### ISSUE THREAD

**Issue by SzBosch** [conan-io/conan#17414]
### What is your question?

Hello together,

the following commands:

conan lock create --update --requires="pkg/1.0.0"  # create conan.lock
conan install --update --requires="pkg/1.0.1"  # conan.lock is used implicitly

gives the error:

ERROR: Requirement 'pkg/1.0.1' not in lockfile

How can we use the lockfile to overrule the requirements version?

Goal: The 2nd command

conan install --update --requires="pkg/1.0.1"  # conan.lock is used implicitly

shall install "pkg/1.0.0".

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @SzBosch 

Thanks for your question.

It is important to understand the ``lockfiles`` as a snapshot of a given dependency graph versions, that allows to reproduce exactly those graph versions, under the same initial conditions, by constraining the possible version ranges that exist. But a lockfile is a constraint mechanism, not a version definition/redefinition mechanism.

So if the first command ``conan lock create`` creates a ``conan.lock`` that contains inside ``pkg/1.0.0``, using that lockfile will effectively constrain existing ranges to that ``pkg/1.0.0``.

But the second command is not providing a range, but instead a fixed ``pkg/1.0.1`` version, that Conan cannot found in the lockfile and then by default it errors, trying to protect against not locked dependencies.

This is better understood if using version ranges as:
- ``conan lock create --requires="pkg/[>=1.0 <2]"`` creates a lockfile with ``pkg/1.0`` in it
- A new package ``pkg/1.1`` exists now
- ``conan install --requires="pkg/[>=1.0 <2]" --lockfile=conan.lock`` will effectively enforce and constrain to use ``pkg/1.0`` instead of ``pkg/1.1`` even if that would be the default.


What you want to achieve can be done by explicitly telling Conan to "relax" the lockfile with:
```
conan install --update --requires="pkg/1.0.1" --lockfile-partial
```
You can even capture the new lockfile with ``--lockfile-out`` that would include the new used version
```
conan install --requires=pkg/1.0.1 --lockfile-partial --lockfile-out=new.lock
```

**Comment by SzBosch** [comment]
Hello @memsharded , thank you for your answer.

I undestood now for a complete freeze the input dependency graph (means: the whole recipe: conanfile.py and all related files) needs also to be frozen, additionally to the the lockfile. Or in other words, both need to be baselined together (e.g. by storing them in the same git repo) to ensure reproduce-ability.

**Comment by memsharded** [comment]
> I undestood now for a complete freeze the input dependency graph (means: the whole recipe: conanfile.py and all related files) needs also to be frozen, additionally to the the lockfile. Or in other words, both need to be baselined together (e.g. by storing them in the same git repo) to ensure reproduce-ability.

Yes, this is also why we keep working on features like the ``conan config install-pkg`` that allows Conan packages to store configuration, and those packages can also be locked in lockfiles, so that allows to achieve reproducibility also over the full configuration. So one repository with one conanfile and a lockfile could achieve configuration and dependencies reproducibility.

**Comment by memsharded** [comment]
Following up on this old ticket.

Now conan learned ``conan config install-pkg`` with the ``conanconfig.yml`` file and also with lockfile capabilities, so it is now way easier to achieve reproducibility over configuration too.

I am closing this question as answered, but feel free to comment or open new tickets for any further question or issue. Thanks for the feedback!
